### DISCUSSION TITLE
PostgresOperator and passing list to WHERE IN statement

### DISCUSSION THREAD

**Discussion by kiranzo** [apache/airflow#34920] [Q&A]
[How-to Guide for PostgresOperator](https://airflow.apache.org/docs/apache-airflow-providers-postgres/stable/operators/postgres_operator_howto_guide.html) does not include examples of how to handle lists of values to pass to WHERE IN statement. 

From the information I found, I'm supposed to render the whole list of values to SQL-compatible format myself. Is there really no way to let Airflow Postgres functionality to do that for me?

Can you please update PostgresOperator documentation with this info as well?

**Comment by Taragolis** [comment]
PostgreSQL operator build in top of `psycopg2`, so it basically use [psycopg2: Passing parameters to SQL queries](https://www.psycopg.org/docs/usage.html#passing-parameters-to-sql-queries) 

So I guess it should be something like

```python
PostgresOperator(
    task_id="awesome_task",
    sql="SELECT * FROM foo WHERE bar in %(awesome_parameter)s;",
    parameters={
        "awesome_parameter": ("spam", "egg")  # Note it should be a tuple
    },
    ...
)
```

**Reply by kiranzo** [reply]
Thanks, trying this now.

How to connect the output of PythonOperator to input of PostgresOperator? I tried returning tuple of ids of rows I want to delete, passing `parameters={'ids': my_python_task.output}` and got `psycopg2.ProgrammingError: can't adapt type 'PlainXComArg'`

Seems like passing info between different types of operators is a huge nuisance.

**Reply by Taragolis** [reply]
This works fine for me

```python
from airflow.decorators import task
from airflow.models.dag import DAG
from airflow.operators.python import PythonOperator
from airflow.providers.common.sql.operators.sql import SQLExecuteQueryOperator
from airflow.utils.timezone import datetime


def python_operator_params():
    return "pg_conn", "aws_default", "foo_bar"


with DAG(
    "discussion_34920",
    start_date=datetime(2023, 10, 1),
    schedule_interval=None,
    catchup=False,
    tags=["discussion", "34920", "SQLExecuteQueryOperator"]
) as dag:

    @task
    def parameters():
        return "pg_conn", "aws_default", "foo_bar"

    task_flow_task = parameters()

    python_operator_task = PythonOperator(
        task_id="python-operator-task",
        python_callable=python_operator_params
    )

    task_sql_op_1 = SQLExecuteQueryOperator(
        task_id="sql-operator-from-task-flow",
        conn_id="pg_conn",
        sql="SELECT id FROM public.connection WHERE conn_id IN %(conns)s;",
        parameters={
            "conns": task_flow_task
        },
    )

    task_sql_op_2 = SQLExecuteQueryOperator(
        task_id="sql-operator-from-classic-operator",
        conn_id="pg_conn",
        sql="SELECT id FROM public.connection WHERE conn_id IN %(conns)s;",
        parameters={
            "conns": python_operator_task.output
        },
    )

    task_flow_task >> task_sql_op_1
    python_operator_task >> task_sql_op_2

```

**Reply by Taragolis** [reply]
I guess you use deprecated `PostgresOperator`, in their definition it overwrites `template_fields` so better to stitch to `SQLExecuteQueryOperator`

If you still need to use `PostgresOperator` then you could monkeypatch them until it not fixed


```python
from airflow.providers.postgres.operators.postgres import PostgresOperator

PostgresOperator.template_fields = ("conn_id", "sql", "parameters")
```

**Reply by Taragolis** [reply]
FYI: #34969

**Reply by kiranzo** [reply]
Thanks, I did everything with PostgresHook for now, but I'll redo simple queries via SQLExecuteQueryOperator when I revisit it next time.
